perun-core/src/main/java/cz/metacentrum/perun/core/entry/ResourcesManagerEntry.java

```diff
@@ -99,6 +99,32 @@ public Resource createResource(PerunSession sess, Resource resource, Vo vo, Faci
 		return getResourcesManagerBl().createResource(sess, resource, vo, facility);
 	}
 
+	public Resource copyResource(PerunSession sess, Resource templateResource, Resource destinationResource, boolean withGroups) throws InternalErrorException, ResourceNotExistsException, FacilityNotExistsException, PrivilegeException, VoNotExistsException, ResourceExistsException {
+		Utils.checkPerunSession(sess);
+		Vo destinationVo = getPerunBl().getVosManagerBl().getVoById(sess, destinationResource.getVoId());
+		Facility destinationFacility = getPerunBl().getFacilitiesManagerBl().getFacilityById(sess, destinationResource.getFacilityId());
+		getPerunBl().getVosManagerBl().checkVoExists(sess, destinationVo);
+		getPerunBl().getFacilitiesManagerBl().checkFacilityExists(sess, destinationFacility);
+
+		Facility templateResourceFacility = getPerunBl().getFacilitiesManagerBl().getFacilityById(sess, templateResource.getFacilityId());
+
+		//check if user is part of template resource's facility
+		if(!AuthzResolver.isAuthorized(sess, Role.VOADMIN, templateResource) &&
+				!AuthzResolver.isAuthorized(sess,Role.FACILITYADMIN, templateResourceFacility)){
+			throw new PrivilegeException(sess, "Problem with template Resource's facility in copyResource.");
+		}
+
+		//check if user is part of destination's facility
+		if (!AuthzResolver.isAuthorized(sess, Role.VOADMIN, destinationVo) &&
+				!AuthzResolver.isAuthorized(sess, Role.FACILITYADMIN, destinationFacility)) {
+			throw new PrivilegeException(sess, "Problem with new Resource's facility in copyResource.");
+		}
+
+		getResourcesManagerBl().checkResourceExists(sess, templateResource);
+
+		return getResourcesManagerBl().copyResource(sess, templateResource, destinationResource, withGroups);
```

Comment:
I am still missing a part of code checking if variable "withGroup" is true or false.
False => we don't need to check if resources are from the same VO and user has VOADMIN role in this VO
True => because we want to copy groups, we need to check user's role in VO (it should be VOADMIN) and also both resources should be from the same VO (throw exception if not)